src/lib/components/TocScrollSpy/index.js

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the Licenspe.
+ */
+
+import {ScrollSpy} from '../ScrollSpy';
+
+/**
+ * Element that adds active classes to its children when the user scrolls to
+ * a heading this child points to.
+ * @extends {ScrollSpy}
+ * @final
+ */
+class TocScrollSpy extends ScrollSpy {
+  constructor() {
+    super();
+    this.selector = this.selector = 'h2[id], h3[id]';
+    this.tocActiveClass = 'w-scroll-spy__active';
+    this.tocVisibleClass = 'w-scroll-spy__visible';
+  }
+
+  scrollSpy(headings) {
```

Comment:
This function overrides the `scrollSpy()` function in the base `ScrollSpy` class, right? But it feel like it behaves very differently—instead of firing `'viewport-entry'` events, for instance, it performs DOM lookups and CSS class manipulations.

Instead of overriding `scrollSpy()`, what if the `TocScrollSpy` class was responsible for registering the listener for `'viewport-entry'` events, and reacted to those events with logic similar to what's in this `scrollSpy()` implementation? That's kind of what I assumed you wanted to accomplish via the refactor.